From 97a9f4ebed670e55cfe0fd29e23486a4889cc8b0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Amadeusz=20S=C5=82awi=C5=84ski?=
 <amadeuszx.slawinski@intel.com>
Date: Mon, 10 Dec 2018 11:08:32 +0100
Subject: [PATCH 242/292] ASoC: Topology: Cleanup dynamically registered dais
 on tplg unload
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We called snd_soc_register_dai in soc_tplg_dai_create, so we need to
remove this DAI when we unload topology.

Change-Id: I8366eca7919660a14f24751c27a9d918a1d0cc2c
Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@intel.com>
Reviewed-on:
---
 sound/soc/soc-topology.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sound/soc/soc-topology.c b/sound/soc/soc-topology.c
index b3b1f37f81d3..a76423fcfcb0 100644
--- a/sound/soc/soc-topology.c
+++ b/sound/soc/soc-topology.c
@@ -502,6 +502,7 @@ static void remove_widget(struct snd_soc_component *comp,
 static void remove_dai(struct snd_soc_component *comp,
 	struct snd_soc_dobj *dobj, int pass)
 {
+	struct snd_soc_dai *dai;
 	struct snd_soc_dai_driver *dai_drv =
 		container_of(dobj, struct snd_soc_dai_driver, dobj);
 	struct snd_soc_dai *dai;
@@ -516,6 +517,14 @@ static void remove_dai(struct snd_soc_component *comp,
 		if (dai->driver == dai_drv)
 			dai->driver = NULL;
 
+	for_each_component_dais(comp, dai) {
+		if (dai->driver == dai_drv) {
+			dai->driver = NULL;
+			snd_soc_unregister_dai(comp, dai);
+			break;
+		}
+	}
+
 	kfree(dai_drv->name);
 	list_del(&dobj->list);
 	kfree(dai_drv);
-- 
2.21.0

